substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java

```diff
@@ -78,24 +90,42 @@ public static UnsignedWord size() {
             throw VMError.shouldNotReachHere("Accessing the physical memory size may require allocation and synchronization");
         }
 
-        if (!isInitialized()) {
-            /*
-             * Multiple threads can race to initialize the cache. This is OK because all of them
-             * will (most-likely) compute the same value.
-             */
-            INITIALIZING.incrementAndGet();
-            try {
-                long memoryLimit = SubstrateOptions.MaxRAM.getValue();
-                if (memoryLimit > 0) {
-                    cachedSize = WordFactory.unsigned(memoryLimit);
-                } else {
-                    memoryLimit = Containers.memoryLimitInBytes();
-                    cachedSize = memoryLimit > 0
-                                    ? WordFactory.unsigned(memoryLimit)
-                                    : ImageSingletons.lookup(PhysicalMemorySupport.class).size();
+        synchronized (INITIALIZING) {
+            if (!isInitialized()) {
+                VMError.guarantee(INITIALIZING.get() <= 1, "Must not initialize twice");
```

Comment:
This seems redundant. Since the code is synchronized on INITIALIZING and `isInitialized()` is `false`, it is guaranteed that `INITIALIZING.get() <= 1` will be true, but even if it wasn't this check would be pretty "racy".